src/classes/HiFiCommunicator.ts

```diff
@@ -305,6 +406,78 @@ export class HiFiCommunicator {
+                  AND IF autoRetryInitialConnection is true
+                  AND IF maxSecondsToSpendRetryingInitialConnection has not yet elapsed
+                  THEN ensure that this._currentHiFiConnectionState stays at CONNECTING
+                  ELSE handle the new state as implemented below
+
+               2. ELSE IF the new state is CONNECTED
+                  THEN handle the new state as implemented below
+                  AND kill the maxSecondsToSpendRetryingInitialConnection timeout if it's set
+
+               3. ELSE IF the new state is other (CONNECTING or RECONNECTING)
+                  AND IF the new state is not the same as this._currentHiFiConnectionState
+                  THEN handle the new state as implemented below
+
+            ii) IF this._currentHiFiConnectionState is currently CONNECTED or RECONNECTING:
+
+               1. IF the new state is DISCONNECTED, FAILED, or UNAVAILABLE
+                  AND IF autoRetryOnDisconnect is true
+                  AND IF maxSecondsToSpendRetryingOnDisconnect has not yet elapsed
+                  THEN set this._currentHiFiConnectionState to RECONNECTING
+                  AND try to reconnect
+                  ELSE handle the new state as implemented below
+
+               2. ELSE IF the new state is CONNECTED
+                  THEN handle the new state as implemented below
+                  AND kill the maxSecondsToSpendRetryingOnDisconnect timeout if it's set
+
+               4. ELSE IF the new state is other (CONNECTING or RECONNECTING)
+                  AND IF the new state is not the same as this._currentHiFiConnectionState
+                  THEN handle the new state as implemented below
+
+         */
+        if (this._currentHiFiConnectionState !== state) {
+            // TODO: If this._currentHiFiConnectionState is "UNAVAILABLE", subsequent changes to "FAILED" or "DISCONNECTED"
+            // should probably not trigger a change (they don't in the current implementation), because "UNAVAIALBLE" is
+            // itself a failure status (for "velvet rope"). However, if the change from "UNAVAILABLE" goes to something else (e.g. "CONNECTING")
+            // then the change _should_ be triggered. Need to think through that a little bit more, and also re-evaluate how that behavior has
+            // changed since the websocket signaling changes were implemented (the UNAVAILABLE state gets communicated the same way, which
+            // is what made its original handling logic so very complicated before RaviSession was updated with the new short-circuit approach).
+            // It's probably easier now (i.e. I think the RaviSession may just stay at UNAVAILABLE, in which case this would all be moot).
+            this._currentHiFiConnectionState = state;
+            if (this.onConnectionStateChanged) {
+                this.onConnectionStateChanged(this._currentHiFiConnectionState);
+            }
```

Comment:
How will the user's onConnectionStateChanged handler interact with our stuff (the TODO). E.g., which gets called first? Can one change the state as seen by the other?

Reply:
My intent was to follow our logic first (including kicking off retries as needed), and then call the user's onConnectionStateChanged handler _only if_ there is an actual change in state after following our logic. 

That would happen regardless of whether the change in state was the result of a change bubbled up from the webrtc session or was due to a 'conscious' decision on our part. For instance, if the underlying webrtc session had been connected, and then it disconnected (which triggered _our_ handler), and then we immediately started retrying, the TODO logic would explicitly change the overall Communicator's state to "RECONNECTING" and trigger the user's handler with "RECONNECTING" as the user-facing state, so that the user could put up a "we're trying to reconnect you" message if they want to. If that reconnect attempt failed and we tried to reconnect again, we wouldn't call the user's handler a second time with the same "RECONNECTING" state. However, if we ultimately timed out the attempts to reconnect, we would then switch to "FAILED" and call the user's handler.

The user's connection handler (or any user code) could hypothetically change HiFiCommunicator._currentHiFiConnectionState directly if they really wanted to, because TypeScript private doesn't prevent JavaScript from messing with things. (Unless we change stuff and use new JS features: https://www.valentinog.com/blog/private/ ) But that's JavaScript for you.